CRM/Sepa/Logic/Queue/Update.php

```diff
@@ -23,121 +23,182 @@
-        break;
+    $bgqueue_enabled = (bool) Civi::settings()->get('enableBackgroundQueue');
 
-      case 'CLOSE':
-        $this->title = ts("Cleaning up ended mandates", array('domain' => 'org.project60.sepa'));
-        break;
+    if (!$bgqueue_enabled) {
+      $runner_title = ts('Updating %1 SEPA Groups', [
+        1        => $mode,
+        'domain' => 'org.project60.sepa',
+      ]);
 
-      case 'UPDATE':
-        $this->title = ts("Process {$this->mode} mandates (%1-%2)",
-          array(1 => $this->offset, 2 => $this->offset+$this->limit, 'domain' => 'org.project60.sepa'));
-        break;
+      $runner = new CRM_Queue_Runner([
+        'errorMode' => CRM_Queue_Runner::ERROR_ABORT,
+        'onEndUrl'  => CRM_Utils_System::url('civicrm/sepa/dashboard', 'status=active'),
+        'queue'     => $queue,
+        'title'     => $runner_title,
+      ]);
 
-      case 'CLEANUP':
-        $this->title = ts("Cleaning up {$this->mode} groups",
-          array(1 => $this->offset, 2 => $this->offset+$this->limit, 'domain' => 'org.project60.sepa'));
-        break;
+      $runner->runAllViaWeb();
+    }
+  }
 
-      case 'FINISH':
-        $this->title = ts("Lock released", array('domain' => 'org.project60.sepa'));
-        break;
+  /**
+   * Execute a task
+   * @param $_ctx CRM_Queue_TaskContext
+   * @param $cmd string
+   * @param $params array
+   * @return bool
+   */
+  public static function run($_ctx, $cmd, $params) {
+    $_ctx->queue->setStatus('active');
 
-      default:
-        $this->title = "Unknown";
-      }
-  }
+    $creditor_id = $params['creditor_id'] ?? NULL;
+    $limit = $params['limit'] ?? NULL;
+    $mode = $params['mode'] ?? NULL;
+    $offset = $params['offset'] ?? NULL;
 
-  public function run($context) {
-    switch ($this->cmd) {
+    switch ($cmd) {
       case 'PREPARE':
         // nothing to do
         break;
 
       case 'CLOSE':
         CRM_Sepa_Logic_Batching::closeEnded();
-        CRM_Sepa_Logic_Settings::renewAsyncLock('sdd_async_update_lock', SDD_UPDATE_RUNNER_BATCH_LOCK_TIMOUT);
+
+        CRM_Sepa_Logic_Settings::renewAsyncLock(
+          'sdd_async_update_lock',
+          SDD_UPDATE_RUNNER_BATCH_LOCK_TIMOUT
+        );
+
         break;
 
       case 'UPDATE':
-        if ($this->mode == 'OOFF') {
-          CRM_Sepa_Logic_Batching::updateOOFF($this->creditor_id, 'now', $this->offset, $this->limit);
+        if ($mode == 'OOFF') {
+          CRM_Sepa_Logic_Batching::updateOOFF($creditor_id, 'now', $offset, $limit);
         } else {
-          CRM_Sepa_Logic_Batching::updateRCUR($this->creditor_id, $this->mode, 'now', $this->offset, $this->limit);
+          CRM_Sepa_Logic_Batching::updateRCUR($creditor_id, $mode, 'now', $offset, $limit);
         }
-        CRM_Sepa_Logic_Settings::renewAsyncLock('sdd_async_update_lock', SDD_UPDATE_RUNNER_BATCH_LOCK_TIMOUT);
+
+        CRM_Sepa_Logic_Settings::renewAsyncLock(
```

Comment:
I'm not sure if it was related to this lock, the one in `CRM_Sepa_Logic_Batching::updateRCUR()` or some other lock, but while running this on dev2 with coworker set to 4 workers, I saw quite a few `GET_LOCK()` calls via MySQL's `show processlist`; my guess is that the locking mechanism is effectively preventing us from processing (`UPDATE`) tasks in parallel. We should probably start by figuring out which steps specifically are using which locks and why. It probably makes sense to have locks that prevent more than one overall batching process from running at the same time, but multiple `UPDATE` tasks running at once should be fine.

Reply:
Fixed in 1cb64a09f3e2aaca7ea02f6880283b63339cb315